iroh-p2p/src/node.rs

```diff
@@ -1147,20 +1172,51 @@ mod tests {
 
             let network_events = p2p.network_events();
             let task = tokio::task::spawn(async move { p2p.run().await.unwrap() });
+
+            let client = client.try_p2p()?;
+
+            let addr = get_addr(client.clone()).await?;
+            let mut dial_addr = addr.clone();
+            dial_addr.push(Protocol::P2p(peer_id.into()));
             Ok(TestRunner {
                 task,
-                client: client.try_p2p()?,
+                client,
                 peer_id,
                 network_events,
+                addr,
+                dial_addr,
             })
         }
     }
 
+    async fn get_addr(client: P2pClient) -> Result<Multiaddr> {
+        loop {
+            tokio::select! {
+                _ = tokio::time::sleep(Duration::from_millis(500)) => {
+                    return Err(anyhow!("timed out before we could get a listening address for the node"));
+                },
+                l = client.listeners() => {
+                    if let Some(a) = l?.get(0) {
+                        return Ok(a.clone());
+                    }
+                }
+            }
+        }
```

Comment:
Doing something with a timeout is often simpler to write with
```
let l = tokio::time::timeout(Duration::from_millis(500), client.listeners()).context("waiting for node listening addr")?;
```

The other thing here is that the timeout applies to each select call, while maybe it should be global.  I'm struggling to find a nice way to do this simply, maybe `futures::stream::unfold` isn't too bad for this.  But it's also possible, and might be easier to follow, to wrap the loop into a separate `async fn` and use that future in the timeout.